### PR TITLE
perf: use `jobs.getQueryResults` to download result sets

### PR DESCRIPTION
**Pull request by tswast**
Replaces `tabledata.list` when `RowIterator` is used for query results.
This likely also fixes a few edge cases around BigQuery scripting jobs.

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Based on https://github.com/googleapis/python-bigquery/pull/361
Towards #362  🦕

### CODE REVIEW

File: google/cloud/bigquery/client.py [hunk 1]
```diff
@@ -3169,6 +3170,77 @@ def list_rows(
...
+
+        row_iterator = RowIterator(
+            client=self,
+            api_request=functools.partial(self._call_api, retry, timeout=timeout),
+            path=f"/projects/{project}/queries/{job_id}",
+            schema=schema,
+            max_results=max_results,
+            page_size=page_size,
+            table=destination,
+            extra_params=params,
```
Comment (shollyman) [inline google/cloud/bigquery/client.py:3242]: My assumption is that location gets plumbed through extra params when relevant?

**Review by shollyman — APPROVED** [review]
Looks good.  If you end up reusing _list_rows_from_query_results for the done polling you may need to consider that explicit project list, but it seems you're being distinct between the job polling and iteration usages here.

File: google/cloud/bigquery/client.py (hunk 1 above)
Comment (tswast) [reply google/cloud/bigquery/client.py:3242]: Good catch. I don't believe we are populating location here.

Comment (tswast) [reply google/cloud/bigquery/client.py:3248]: populating in latest commit

**Pull request merged**